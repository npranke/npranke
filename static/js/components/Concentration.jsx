import React from 'react'

import { getLocationPageTitle } from '../utils'
import Picture0 from '../../img/concentration-22332-pink-plateau.jpg'
import Picture1 from '../../img/concentration-22512-blue-center.jpg'
import Picture2 from '../../img/concentration-22585-teal-ripples.jpg'
import Picture3 from '../../img/concentration-22594-burgundy-spot.jpg'
import Picture4 from '../../img/concentration-23056-mauve-mound.jpg'
import Picture5 from '../../img/concentration-23080-lavendar-corner.jpg'
import Picture6 from '../../img/concentration-23103-light-pool.jpg'
import Picture7 from '../../img/concentration-23183-brown-variations.jpg'
import Picture8 from '../../img/concentration-23184-purple-surface.jpg'
import Picture9 from '../../img/concentration-23231-bright-swoop.jpg'
import Picture10 from '../../img/concentration-23233-dark-ridges.jpg'
import Picture11 from '../../img/concentration-23239-neon-plains.jpg'

class Concentration extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            isPortrait: window.matchMedia(
                '(orientation: portrait)',
            ).matches,
        }

        this.updateIsPortrait = this.updateIsPortrait.bind(this)
    }

    componentDidMount() {
        document.title = getLocationPageTitle('concentration')

        window.matchMedia('(orientation: portrait)').addEventListener(
            'change',
            this.updateIsPortrait,
            false,
        )
    }

    componentWillUnmount() {
        window.matchMedia('(orientation: portrait)').removeEventListener(
            'change',
            this.updateIsPortrait,
            false,
        )
    }

    updateIsPortrait() {
        this.setState({
            isPortrait: window.matchMedia(
                '(orientation: portrait)',
            ).matches,
        })
    }

    render() {
        const board = this.state.isPortrait
            ? (
                <div className="board">
                    <div className="board-row-wrapper">
                        <div className="board-row">
                            <img
                                src={ Picture0 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture0 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture1 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture1 }
                                className="picture"
                                alt="mars"
                            />
                        </div>
                    </div>
                    <div className="board-row-wrapper">
                        <div className="board-row">
                            <img
                                src={ Picture2 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture2 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture3 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture3 }
                                className="picture"
                                alt="mars"
                            />
                        </div>
                    </div>
                    <div className="board-row-wrapper">
                        <div className="board-row">
                            <img
                                src={ Picture4 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture4 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture5 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture5 }
                                className="picture"
                                alt="mars"
                            />
                        </div>
                    </div>
                    <div className="board-row-wrapper">
                        <div className="board-row">
                            <img
                                src={ Picture6 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture6 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture7 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture7 }
                                className="picture"
                                alt="mars"
                            />
                        </div>
                    </div>
                    <div className="board-row-wrapper">
                        <div className="board-row">
                            <img
                                src={ Picture8 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture8 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture9 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture9 }
                                className="picture"
                                alt="mars"
                            />
                        </div>
                    </div>
                    <div className="board-row-wrapper">
                        <div className="board-row">
                            <img
                                src={ Picture10 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture10 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture11 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture11 }
                                className="picture"
                                alt="mars"
                            />
                        </div>
                    </div>
                </div>
            ) : (
                <div className="board">
                    <div className="board-row-wrapper">
                        <div className="board-row">
                            <img
                                src={ Picture0 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture0 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture1 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture1 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture2 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture2 }
                                className="picture"
                                alt="mars"
                            />
                        </div>
                    </div>
                    <div className="board-row-wrapper">
                        <div className="board-row">
                            <img
                                src={ Picture3 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture3 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture4 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture4 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture5 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture5 }
                                className="picture"
                                alt="mars"
                            />
                        </div>
                    </div>
                    <div className="board-row-wrapper">
                        <div className="board-row">
                            <img
                                src={ Picture6 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture6 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture7 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture7 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture8 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture8 }
                                className="picture"
                                alt="mars"
                            />
                        </div>
                    </div>
                    <div className="board-row-wrapper">
                        <div className="board-row">
                            <img
                                src={ Picture9 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture9 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture10 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture10 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture11 }
                                className="picture"
                                alt="mars"
                            />
                            <img
                                src={ Picture11 }
                                className="picture"
                                alt="mars"
                            />
                        </div>
                    </div>
                </div>
            )

        return (
            <main className="workbook workbook-concentration">
                <div className="concentration">
                    <div className="concentration-inner">
                        <div className="matches-wrapper">
                            <div className="matches">
                                <div className="matches-header">
                                    <div className="left-text">matches: 4</div>
                                    <div className="right-text">
                                        time: 00:00:00
                                    </div>
                                </div>
                                <div className="matches-pictures">
                                    <img
                                        src={ Picture0 }
                                        className="picture"
                                        alt="mars"
                                        id="first-match"
                                    />
                                    <img
                                        src={ Picture1 }
                                        className="picture"
                                        alt="mars"
                                    />
                                    <img
                                        src={ Picture2 }
                                        className="picture"
                                        alt="mars"
                                    />
                                    <img
                                        src={ Picture3 }
                                        className="picture"
                                        alt="mars"
                                    />
                                    <img
                                        src={ Picture4 }
                                        className="picture"
                                        alt="mars"
                                    />
                                    <img
                                        src={ Picture5 }
                                        className="picture"
                                        alt="mars"
                                    />
                                    <img
                                        src={ Picture6 }
                                        className="picture"
                                        alt="mars"
                                    />
                                    <img
                                        src={ Picture7 }
                                        className="picture"
                                        alt="mars"
                                    />
                                    <img
                                        src={ Picture8 }
                                        className="picture"
                                        alt="mars"
                                    />
                                    <img
                                        src={ Picture9 }
                                        className="picture"
                                        alt="mars"
                                    />
                                    <img
                                        src={ Picture10 }
                                        className="picture"
                                        alt="mars"
                                    />
                                    <img
                                        src={ Picture11 }
                                        className="picture"
                                        alt="mars"
                                    />
                                </div>
                            </div>
                        </div>
                        <div className="board-wrapper">
                            { board }
                        </div>
                    </div>
                </div>
            </main>
        )
    }
}

export default Concentration
