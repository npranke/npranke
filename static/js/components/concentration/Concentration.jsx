import PropTypes from 'prop-types'
import React from 'react'

import ConcentrationBoard from '@components/concentration/ConcentrationBoard'
import ConcentrationMatches from
    '@components/concentration/ConcentrationMatches'
import withPortraitListener from '@components/hoc/PortraitListener'

import { PICTURE_BACK, PICTURE_SOURCES } from '@constants/concentration'

import { getSendEventHandler, sendEvent } from '@utils'

export class Concentration extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            isTimeRunning: false,
            turns: 0,
            internalMatches: [],
            displayedMatches: [],
            first: { id: null, pictureid: null },
            second: { id: null, pictureid: null },
        }

        this.pictureClickHandler = this.pictureClickHandler.bind(this)
        this.pictureKeyUpHandler = this.pictureKeyUpHandler.bind(this)

        this.matchesPictures = Object.keys(PICTURE_SOURCES).reduce(
            (pictureAccumulator, pictureid) => {
                const mroPIA = `NASA/JPL Mars Reconnaissance Orbiter PIA${
                    PICTURE_SOURCES[pictureid].sourceid
                }`

                pictureAccumulator[`picture${pictureid}`] = {
                    image: (
                        <a
                            onClick={ getSendEventHandler(
                                'concentration',
                                'navigate',
                                `PIA${PICTURE_SOURCES[pictureid].sourceid}`,
                            ) }
                            href={
                                'https://www.jpl.nasa.gov/spaceimages/'
                                + 'details.php?id='
                                + `PIA${PICTURE_SOURCES[pictureid].sourceid}`
                            }
                            target="_blank"
                            rel="noreferrer noopener external"
                            aria-label="NASA/JPL Space Images"
                        >
                            <img
                                src={ PICTURE_SOURCES[pictureid].source }
                                className="picture"
                                alt={ mroPIA }
                                title={ mroPIA }
                            />
                        </a>
                    ),
                }

                return pictureAccumulator
            },
            {},
        )

        this.boardPictures = Object.keys(PICTURE_SOURCES).reduce(
            (pictureAccumulator, pictureid) => {
                const mroPIA = `NASA/JPL Mars Reconnaissance Orbiter PIA${
                    PICTURE_SOURCES[pictureid].sourceid
                }`
                const merPIA = 'NASA/JPL Mars Exploration Rover PIA20285'

                pictureAccumulator[`picture${pictureid}-a`] = {
                    id: `${pictureid}-a`,
                    pictureid: `${pictureid}`,
                    back: (
                        <img
                            onClick={ this.pictureClickHandler }
                            onKeyUp={ this.pictureKeyUpHandler }
                            role="button"
                            tabIndex="0"
                            data-id={ `${pictureid}-a` }
                            data-pictureid={ pictureid }
                            src={ PICTURE_BACK }
                            className="picture picture-back"
                            alt={ merPIA }
                        />
                    ),
                    front: (
                        <img
                            src={ PICTURE_SOURCES[pictureid].source }
                            className="picture picture-front"
                            alt={ mroPIA }
                        />
                    ),
                    matched: (
                        <img
                            src={ PICTURE_SOURCES[pictureid].source }
                            className="picture picture-matched"
                            alt={ mroPIA }
                        />
                    ),
                }

                pictureAccumulator[`picture${pictureid}-b`] = {
                    id: `${pictureid}-b`,
                    pictureid: `${pictureid}`,
                    back: (
                        <img
                            onClick={ this.pictureClickHandler }
                            onKeyUp={ this.pictureKeyUpHandler }
                            role="button"
                            tabIndex="0"
                            data-id={ `${pictureid}-b` }
                            data-pictureid={ pictureid }
                            src={ PICTURE_BACK }
                            className="picture picture-back"
                            alt={ merPIA }
                        />
                    ),
                    front: (
                        <img
                            src={ PICTURE_SOURCES[pictureid].source }
                            className="picture picture-front"
                            alt={ mroPIA }
                        />
                    ),
                    matched: (
                        <img
                            src={ PICTURE_SOURCES[pictureid].source }
                            className="picture picture-matched"
                            alt={ mroPIA }
                        />
                    ),
                }

                return pictureAccumulator
            },
            {},
        )

        this.boardOrder = Object.keys(
            this.boardPictures,
        ).map((pictureName) => {
            return { sortValue: Math.random(), pictureName }
        }).sort((pictureWrapperA, pictureWrapperB) => {
            return pictureWrapperA.sortValue - pictureWrapperB.sortValue
        }).map((pictureWrapper) => {
            return pictureWrapper.pictureName
        })
    }

    componentWillUnmount() {
        clearTimeout(this.pictureTimeout)
    }

    pictureClickHandler(event) {
        const { id, pictureid } = event.currentTarget.dataset

        sendEvent('concentration', 'click', 'picture')

        this.setState((state) => {
            if (state.first.id === null) {
                return {
                    first: { id, pictureid },
                    displayedMatches: state.internalMatches,
                    isTimeRunning: true,
                }
            } else if (
                state.second.id === null && state.first.pictureid === pictureid
            ) {
                sendEvent('concentration', 'find', 'match')

                const isLastMatch = state.internalMatches.length === 11

                if (isLastMatch) {
                    sendEvent(
                        'concentration',
                        'complete',
                        'matches',
                        state.turns + 1,
                    )
                }

                this.pictureTimeout = setTimeout(() => {
                    this.setState({
                        first: { id: null, pictureid: null },
                        second: { id: null, pictureid: null },
                        displayedMatches: state.displayedMatches.concat(
                            [pictureid],
                        ),
                        isTimeRunning: !isLastMatch,
                    })
                }, 1000)

                return {
                    turns: state.turns + 1,
                    second: { id, pictureid },
                    internalMatches: state.internalMatches.concat([pictureid]),
                    displayedMatches: state.internalMatches,
                    isTimeRunning: !isLastMatch,
                }
            } else if (
                state.second.id === null && state.first.pictureid !== pictureid
            ) {
                this.pictureTimeout = setTimeout(() => {
                    this.setState({
                        first: { id: null, pictureid: null },
                        second: { id: null, pictureid: null },
                        displayedMatches: state.internalMatches,
                        isTimeRunning: true,
                    })
                }, 2000)

                return {
                    turns: state.turns + 1,
                    second: { id, pictureid },
                    displayedMatches: state.internalMatches,
                    isTimeRunning: true,
                }
            } else if (state.first.id && state.second.id) {
                clearTimeout(this.pictureTimeout)

                return {
                    first: { id, pictureid },
                    second: { id: null, pictureid: null },
                    displayedMatches: state.internalMatches,
                    isTimeRunning: true,
                }
            }

            return {
                displayedMatches: state.internalMatches,
                isTimeRunning: true,
            }
        })
    }

    pictureKeyUpHandler(event) {
        if (event.key === 'Enter' || event.key === ' ') {
            this.pictureClickHandler(event)
        }
    }

    render() {
        return (
            <div className="concentration">
                <div className="matches-wrapper">
                    <ConcentrationMatches
                        isTimeRunning={ this.state.isTimeRunning }
                        pictures={ this.matchesPictures }
                        matches={ this.state.displayedMatches }
                        turns={ this.state.turns }
                    />
                </div>
                <div className="board-wrapper">
                    <ConcentrationBoard
                        pictures={ this.boardPictures }
                        boardOrder={ this.boardOrder }
                        isPortrait={ this.props.isPortrait }
                        first={ this.state.first }
                        second={ this.state.second }
                        matches={ this.state.internalMatches }
                    />
                </div>
            </div>
        )
    }
}

Concentration.defaultProps = {
    isPortrait: false,
}

Concentration.propTypes = {
    isPortrait: PropTypes.bool,
}

export default withPortraitListener(Concentration)
