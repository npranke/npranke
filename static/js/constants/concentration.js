import Picture0 from '@img/concentration-22332-pink-plateau.jpg'
import Picture1 from '@img/concentration-22512-blue-center.jpg'
import Picture2 from '@img/concentration-22585-teal-ripples.jpg'
import Picture3 from '@img/concentration-22594-burgundy-spot.jpg'
import Picture4 from '@img/concentration-23056-mauve-mound.jpg'
import Picture5 from '@img/concentration-23080-lavendar-corner.jpg'
import Picture6 from '@img/concentration-23103-light-pool.jpg'
import Picture7 from '@img/concentration-23183-brown-variations.jpg'
import Picture8 from '@img/concentration-23184-purple-surface.jpg'
import Picture9 from '@img/concentration-23231-bright-swoop.jpg'
import Picture10 from '@img/concentration-23233-dark-ridges.jpg'
import Picture11 from '@img/concentration-23239-neon-plains.jpg'
import PictureBack from '@img/concentration-20285-opportunity-rover.jpg'

export const PICTURE_BACK = PictureBack

export const PICTURE_SOURCES = {
    0: { source: Picture0, sourceid: 22332 },
    1: { source: Picture1, sourceid: 22512 },
    2: { source: Picture2, sourceid: 22585 },
    3: { source: Picture3, sourceid: 22594 },
    4: { source: Picture4, sourceid: 23056 },
    5: { source: Picture5, sourceid: 23080 },
    6: { source: Picture6, sourceid: 23103 },
    7: { source: Picture7, sourceid: 23183 },
    8: { source: Picture8, sourceid: 23184 },
    9: { source: Picture9, sourceid: 23231 },
    10: { source: Picture10, sourceid: 23233 },
    11: { source: Picture11, sourceid: 23239 },
}
