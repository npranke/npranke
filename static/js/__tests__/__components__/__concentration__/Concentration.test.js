import Adapter from 'enzyme-adapter-react-16'
import Enzyme, { mount, shallow } from 'enzyme'
import React from 'react'

import { Concentration } from '@components/concentration/Concentration'

import { getSendEventHandler, sendEvent } from '@utils'

jest.mock('@utils')

Enzyme.configure({ adapter: new Adapter() })

describe('Concentration', () => {
    test('has matches section', () => {
        const concentration = shallow(<Concentration />)

        expect(
            concentration.find(
                '.concentration .matches-wrapper ConcentrationMatches',
            ).exists(),
        ).toBe(true)
    })

    test('has board section', () => {
        const concentration = shallow(<Concentration />)

        expect(
            concentration.find(
                '.concentration .board-wrapper ConcentrationBoard',
            ).exists(),
        ).toBe(true)
    })

    test('matchesPictures has 12 pictures', () => {
        const concentration = shallow(<Concentration />)

        expect(
            Object.keys(concentration.instance().matchesPictures),
        ).toHaveLength(12)
    })

    test('boardPictures has 24 pictures', () => {
        const concentration = shallow(<Concentration />)

        expect(
            Object.keys(concentration.instance().boardPictures),
        ).toHaveLength(24)
    })

    test('boardOrder has 24 picture names', () => {
        const concentration = shallow(<Concentration />)

        expect(
            concentration.instance().boardOrder,
        ).toHaveLength(24)
    })

    test('boardOrder is shuffled', () => {
        const concentration = shallow(<Concentration />)

        const exactOrder = Object.keys(
            concentration.instance().boardPictures,
        )

        let exactOrderMatches = 0
        concentration.instance().boardOrder.forEach((pictureName, index) => {
            if (pictureName === exactOrder[index]) {
                exactOrderMatches += 1
            }
        })

        expect(
            exactOrderMatches,
        ).toBeLessThan(24)
    })

    test('time starts running when a picture is clicked', () => {
        const concentration = mount(<Concentration />)

        concentration.find('.picture-back').first().simulate('click')

        expect(
            concentration.state('isTimeRunning'),
        ).toBe(true)
    })

    describe('sending concentration events', () => {
        afterEach(() => {
            getSendEventHandler.mockReset()
            sendEvent.mockReset()
        })

        test('when picture is clicked', () => {
            const concentration = mount(<Concentration />)

            concentration.find('.picture-back').first().simulate('click')

            expect(
                sendEvent,
            ).toHaveBeenNthCalledWith(
                1,
                'concentration',
                'click',
                'picture',
            )
        })

        test('when match is clicked', () => {
            const concentration = mount(<Concentration />)

            const pictureid = concentration.find(
                '.picture-back',
            ).first().props()['data-pictureid']

            concentration.setState({
                first: {
                    id: `${pictureid}-x`,
                    pictureid,
                },
            })

            concentration.find('.picture-back').first().simulate('click')

            expect(
                sendEvent,
            ).toHaveBeenNthCalledWith(
                2,
                'concentration',
                'find',
                'match',
            )
        })

        test('when matches picture is clicked', () => {
            mount(<Concentration />)

            expect(
                getSendEventHandler,
            ).toHaveBeenCalledTimes(12)
            expect(
                getSendEventHandler,
            ).toHaveBeenNthCalledWith(
                1,
                'concentration',
                'navigate',
                'PIA22332',
            )
        })
    })

    describe('pictureClickHandler()', () => {
        test('sets first when first is null', () => {
            const concentration = shallow(<Concentration />)

            concentration.instance().pictureClickHandler({
                currentTarget: {
                    dataset: {
                        id: '2-b',
                        pictureid: '2',
                    },
                },
            })

            expect(
                concentration.state('first'),
            ).toEqual({ id: '2-b', pictureid: '2' })
        })

        test('sets second and internalMatches when match', () => {
            const concentration = shallow(<Concentration />)

            concentration.setState({
                first: { id: '2-b', pictureid: '2' },
            })

            concentration.instance().pictureClickHandler({
                currentTarget: {
                    dataset: {
                        id: '2-a',
                        pictureid: '2',
                    },
                },
            })

            expect(
                concentration.state('second'),
            ).toEqual({ id: '2-a', pictureid: '2' })
            expect(
                concentration.state('internalMatches'),
            ).toEqual(['2'])
        })

        test('sets pictureTimeout when match', () => {
            const concentration = shallow(<Concentration />)

            concentration.setState({
                first: { id: '9-a', pictureid: '9' },
            })

            concentration.instance().pictureClickHandler({
                currentTarget: {
                    dataset: {
                        id: '9-b',
                        pictureid: '9',
                    },
                },
            })

            expect(
                concentration.instance().pictureTimeout,
            ).toBeDefined()
        })

        test('pictureTimeout resets first and second when match', () => {
            jest.useFakeTimers()

            const concentration = shallow(<Concentration />)

            concentration.setState({
                first: { id: '6-a', pictureid: '6' },
            })

            concentration.instance().pictureClickHandler({
                currentTarget: {
                    dataset: {
                        id: '6-b',
                        pictureid: '6',
                    },
                },
            })

            jest.advanceTimersByTime(1000)

            expect(
                concentration.state('first'),
            ).toEqual({ id: null, pictureid: null })
            expect(
                concentration.state('second'),
            ).toEqual({ id: null, pictureid: null })
        })

        test('increments turns when match', () => {
            const concentration = shallow(<Concentration />)

            concentration.setState({
                first: { id: '5-a', pictureid: '5' },
                turns: 3,
            })

            concentration.instance().pictureClickHandler({
                currentTarget: {
                    dataset: {
                        id: '5-b',
                        pictureid: '5',
                    },
                },
            })

            expect(
                concentration.state('turns'),
            ).toEqual(4)
        })

        test('pictureTimeout sets displayedMatches when match', () => {
            jest.useFakeTimers()

            const concentration = shallow(<Concentration />)

            concentration.setState({
                first: { id: '2-b', pictureid: '2' },
                displayedMatches: ['5', '9'],
            })

            concentration.instance().pictureClickHandler({
                currentTarget: {
                    dataset: {
                        id: '2-a',
                        pictureid: '2',
                    },
                },
            })

            jest.advanceTimersByTime(1000)

            expect(
                concentration.state('displayedMatches'),
            ).toEqual(['5', '9', '2'])
        })

        test('sets second and not internalMatches when not match', () => {
            const concentration = shallow(<Concentration />)

            concentration.setState({
                first: { id: '11-a', pictureid: '11' },
            })

            concentration.instance().pictureClickHandler({
                currentTarget: {
                    dataset: {
                        id: '5-a',
                        pictureid: '5',
                    },
                },
            })

            expect(
                concentration.state('second'),
            ).toEqual({ id: '5-a', pictureid: '5' })
            expect(
                concentration.state('internalMatches'),
            ).toEqual([])
        })

        test('sets pictureTimeout when not match', () => {
            const concentration = shallow(<Concentration />)

            concentration.setState({
                first: { id: '8-b', pictureid: '8' },
            })

            concentration.instance().pictureClickHandler({
                currentTarget: {
                    dataset: {
                        id: '10-b',
                        pictureid: '10',
                    },
                },
            })

            expect(
                concentration.instance().pictureTimeout,
            ).toBeDefined()
        })

        test('pictureTimeout resets first and second when not match', () => {
            jest.useFakeTimers()

            const concentration = shallow(<Concentration />)

            concentration.setState({
                first: { id: '10-a', pictureid: '10' },
            })

            concentration.instance().pictureClickHandler({
                currentTarget: {
                    dataset: {
                        id: '11-b',
                        pictureid: '11',
                    },
                },
            })

            jest.advanceTimersByTime(2000)

            expect(
                concentration.state('first'),
            ).toEqual({ id: null, pictureid: null })
            expect(
                concentration.state('second'),
            ).toEqual({ id: null, pictureid: null })
        })

        test('pictureTimeout sets displayedMatches when not match', () => {
            jest.useFakeTimers()

            const concentration = shallow(<Concentration />)

            concentration.setState({
                first: { id: '5-b', pictureid: '5' },
                internalMatches: ['4', '11'],
            })

            concentration.instance().pictureClickHandler({
                currentTarget: {
                    dataset: {
                        id: '6-a',
                        pictureid: '6',
                    },
                },
            })

            jest.advanceTimersByTime(2000)

            expect(
                concentration.state('displayedMatches'),
            ).toEqual(['4', '11'])
        })

        test('increments turns when not match', () => {
            const concentration = shallow(<Concentration />)

            concentration.setState({
                first: { id: '5-b', pictureid: '5' },
                turns: 5,
            })

            concentration.instance().pictureClickHandler({
                currentTarget: {
                    dataset: {
                        id: '7-a',
                        pictureid: '7',
                    },
                },
            })

            expect(
                concentration.state('turns'),
            ).toEqual(6)
        })

        test('sets first and resets second when new picture', () => {
            const concentration = shallow(<Concentration />)

            concentration.setState({
                first: { id: '3-b', pictureid: '3' },
                second: { id: '7-b', pictureid: '7' },
            })

            concentration.instance().pictureClickHandler({
                currentTarget: {
                    dataset: {
                        id: '0-a',
                        pictureid: '0',
                    },
                },
            })

            expect(
                concentration.state('first'),
            ).toEqual({ id: '0-a', pictureid: '0' })
            expect(
                concentration.state('second'),
            ).toEqual({ id: null, pictureid: null })
        })


        test('clears pictureTimeout when new picture', () => {
            const concentration = shallow(<Concentration />)

            concentration.setState({
                first: { id: '3-a', pictureid: '3' },
                second: { id: '3-b', pictureid: '3' },
            })

            concentration.instance().pictureClickHandler({
                currentTarget: {
                    dataset: {
                        id: '0-b',
                        pictureid: '0',
                    },
                },
            })

            expect(
                concentration.instance().pictureTimeout,
            ).toBeUndefined()
        })

        test('sets displayedMatches and isTimeRunning', () => {
            const concentration = shallow(<Concentration />)

            concentration.setState({
                internalMatches: Array.from(
                    { length: 4 },
                    (value, integer) => {
                        return `${integer}`
                    },
                ),
            })

            concentration.instance().pictureClickHandler({
                currentTarget: {
                    dataset: {
                        id: '7-a',
                        pictureid: '7',
                    },
                },
            })

            expect(
                concentration.state('displayedMatches'),
            ).toEqual(['0', '1', '2', '3'])
            expect(
                concentration.state('isTimeRunning'),
            ).toBe(true)
        })
    })

    describe('pictureKeyUpHandler()', () => {
        test('calls pictureClickHandler() on enter', () => {
            const concentration = mount(<Concentration />)

            concentration.instance().pictureClickHandler = jest.fn()

            concentration.find('.picture-back').first().simulate(
                'keyup',
                { key: 'Enter' },
            )

            expect(
                concentration.instance().pictureClickHandler,
            ).toHaveBeenCalled()
        })

        test('calls pictureClickHandler() on spacebar', () => {
            const concentration = mount(<Concentration />)

            concentration.instance().pictureClickHandler = jest.fn()

            concentration.find('.picture-back').first().simulate(
                'keyup',
                { key: ' ' },
            )

            expect(
                concentration.instance().pictureClickHandler,
            ).toHaveBeenCalled()
        })
    })

    describe('updateTime()', () => {
        test('stops time running when matches is 12', () => {
            const concentration = shallow(<Concentration />)

            concentration.setState({
                isTimeRunning: true,
                internalMatches: Array.from(
                    { length: 12 },
                    (value, integer) => {
                        return `${integer}`
                    },
                ),
            })

            concentration.instance().updateTime()

            expect(
                concentration.state('isTimeRunning'),
            ).toBe(false)
        })

        test('formats time when centiseconds is less than 10', () => {
            const concentration = shallow(<Concentration />)

            concentration.setState({ isTimeRunning: true, centiseconds: '5' })

            concentration.instance().updateTime()

            expect(
                concentration.state('centiseconds'),
            ).toEqual('06')
        })

        test('formats time when seconds is less than 10', () => {
            const concentration = shallow(<Concentration />)

            concentration.setState({ isTimeRunning: true, seconds: '8' })

            concentration.instance().updateTime()

            expect(
                concentration.state('seconds'),
            ).toEqual('08')
        })

        test('formats time when minutes is less than 10', () => {
            const concentration = shallow(<Concentration />)

            concentration.setState({ isTimeRunning: true, minutes: '2' })

            concentration.instance().updateTime()

            expect(
                concentration.state('minutes'),
            ).toEqual('02')
        })

        test('formats time when centiseconds is over 100', () => {
            const concentration = shallow(<Concentration />)

            concentration.setState({
                isTimeRunning: true,
                centiseconds: '103',
            })

            concentration.instance().updateTime()

            expect(
                concentration.state('centiseconds'),
            ).toEqual('04')
            expect(
                concentration.state('seconds'),
            ).toEqual('01')
        })

        test('formats time when seconds is over 60', () => {
            const concentration = shallow(<Concentration />)

            concentration.setState({ isTimeRunning: true, seconds: '61' })

            concentration.instance().updateTime()

            expect(
                concentration.state('seconds'),
            ).toEqual('01')
            expect(
                concentration.state('minutes'),
            ).toEqual('01')
        })
    })
})

describe('Concentration snapshot', () => {
    test('matches snapshot', () => {
        const concentration = shallow(<Concentration />)

        concentration.instance().boardOrder = [
            'picture0-a',
            'picture0-b',
            'picture1-a',
            'picture1-b',
            'picture2-a',
            'picture2-b',
            'picture3-a',
            'picture3-b',
            'picture4-a',
            'picture4-b',
            'picture5-a',
            'picture5-b',
            'picture6-a',
            'picture6-b',
            'picture7-a',
            'picture7-b',
            'picture8-a',
            'picture8-b',
            'picture9-a',
            'picture9-b',
            'picture10-a',
            'picture10-b',
            'picture11-a',
            'picture11-b',
        ]

        expect(concentration.instance().render()).toMatchSnapshot()
    })
})
