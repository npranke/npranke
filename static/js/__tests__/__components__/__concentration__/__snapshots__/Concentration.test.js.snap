// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Concentration snapshot matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="concentration"
  >
    <div
      class="matches-wrapper"
    >
      <div
        class="matches"
      >
        <div
          class="matches-header"
        >
          <div
            class="left-text"
          >
            <span
              class="matches-text"
            >
              <span
                class="matches-portion"
              >
                matches: 
              </span>
              <span
                class="matches-portion number"
              >
                00
              </span>
            </span>
            <span
              class="turns-text"
            >
              <span
                class="turns-portion"
              >
                turns: 
              </span>
              <span
                class="turns-portion number"
              >
                00
              </span>
            </span>
          </div>
          <div
            class="right-text"
          >
            <span
              class="time-text"
            >
              <span
                class="time-portion"
              >
                time: 
              </span>
              <span
                class="time-portion number"
                role="timer"
              >
                00
              </span>
              <span
                class="time-portion"
              >
                 : 
              </span>
              <span
                class="time-portion number"
                role="timer"
              >
                00
              </span>
              <span
                class="time-portion"
              >
                 : 
              </span>
              <span
                class="time-portion number"
                role="timer"
              >
                00
              </span>
            </span>
          </div>
        </div>
        <div
          class="matches-pictures"
        />
      </div>
    </div>
    <div
      class="board-wrapper"
    >
      <div
        class="board"
      >
        <div
          class="board-row-wrapper"
        >
          <div
            class="board-row"
          />
        </div>
        <div
          class="board-row-wrapper"
        >
          <div
            class="board-row"
          />
        </div>
        <div
          class="board-row-wrapper"
        >
          <div
            class="board-row"
          />
        </div>
        <div
          class="board-row-wrapper"
        >
          <div
            class="board-row"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
